File: src/csharp/Problems/Preorder.cs
//https://leetcode.com/problems/n-ary-tree-preorder-traversal/

namespace LeetCode.Problems;

public sealed class Preorder : ProblemBase
{
    [Theory]
    [ClassData(typeof(Preorder))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it =>
               {
                   var node = new Node(1);
                   node.children!.Add(new Node(3, new Node(5), new Node(6)));
                   node.children.Add(new Node(2));
                   node.children.Add(new Node(4));

                   it.Param(node).ResultArray("[1,3,5,6,2,4]");
               });

    private IList<int> Solution(Node root)
    {
        if (root == null)
        {
            return Array.Empty<int>();
        }

        var result = new List<int>();
        var stack = new Stack<Node>();
        stack.Push(root);
        while (stack.Count > 0)
        {
            var current = stack.Pop();
            result.Add(current.val);

            for (var i = current.children?.Count - 1; i >= 0; i--)
            {
                stack.Push(current.children![i.Value]);
            }
        }

        return result;
    }

    private IList<int> Solution1(Node root)
    {
        if (root == null)
        {
            return Array.Empty<int>();
        }

        return new int[] { root.val }.Concat(root.children?.SelectMany(Solution1) ?? Array.Empty<int>()).ToArray();
    }
}

File: src/python/problems/Preorder.py

from models.node import Node

class Preorder:
    def Solution(self, root: 'Node') -> list[int]:
        result = []
        stack = []
        stack.append(root)
        while len(stack) > 0:
            current = stack.pop()
            result.append(current.val)

            for i in range(len(current.children) - 1, 0, -1):
                stack.append(current.children[i])

        return result


Preorder().Solution(Node(1, [Node(2), Node(3, [Node(6), Node(7, [Node(11)])])]))
